java/vector/src/main/codegen/templates/ComplexWriters.java

```diff
@@ -113,19 +114,59 @@ public void write(Nullable${minor.class}Holder h) {
     vector.setSafe(idx(), 1<#list fields as field><#if field.include!true >, ${field.name}</#if></#list>);
     vector.setValueCount(idx()+1);
   }
+  </#if>
 
-  <#if minor.class == "Decimal" ||
-       minor.class == "VarChar">
+  <#if minor.class == "VarChar">
   public void write${minor.class}(${friendlyType} value) {
     vector.setSafe(idx(), value);
     vector.setValueCount(idx()+1);
   }
   </#if>
 
   <#if minor.class == "Decimal">
-  public void writeBigEndianBytesToDecimal(byte[] value) {
+
+  public void write(DecimalHolder h){
+    DecimalUtility.checkPrecisionAndScale(h.precision, h.scale, vector.getPrecision(), vector.getScale());
+    vector.setSafe(idx(), h);
+    vector.setValueCount(idx() + 1);
+  }
+
+  public void write(NullableDecimalHolder h){
+    if (h.isSet == 1) {
+      DecimalUtility.checkPrecisionAndScale(h.precision, h.scale, vector.getPrecision(), vector.getScale());
+    }
+    vector.setSafe(idx(), h);
+    vector.setValueCount(idx() + 1);
+  }
+
+  public void writeDecimal(int start, ArrowBuf buffer){
+    vector.setSafe(idx(), 1, start, buffer);
+    vector.setValueCount(idx() + 1);
+  }
+
+  public void writeDecimal(int start, ArrowBuf buffer, ArrowType arrowType){
+    DecimalUtility.checkPrecisionAndScale(((ArrowType.Decimal) arrowType).getPrecision(),
+      ((ArrowType.Decimal) arrowType).getScale(), vector.getPrecision(), vector.getScale());
+    vector.setSafe(idx(), 1, start, buffer);
+    vector.setValueCount(idx() + 1);
+  }
+
+  public void writeDecimal(BigDecimal value){
+    // vector.setSafe already does precision and scale checking
+    vector.setSafe(idx(), value);
+    vector.setValueCount(idx() + 1);
+  }
+
+  public void writeBigEndianBytesToDecimal(byte[] value, ArrowType arrowType){
+    DecimalUtility.checkPrecisionAndScale(((ArrowType.Decimal) arrowType).getPrecision(),
+        ((ArrowType.Decimal) arrowType).getScale(), vector.getPrecision(), vector.getScale());
     vector.setBigEndianSafe(idx(), value);
-    vector.setValueCount(idx()+1);
+    vector.setValueCount(idx() + 1);
+  }
+
+  public void writeBigEndianBytesToDecimal(byte[] value){
+    vector.setBigEndianSafe(idx(), value);
```

Comment:
do we want to support this method..

Reply:
This method was already there (code generated), so I didn't remove it to not break anything. Marked the method as deprecated in the interface.